flink-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/KafkaConsumerThread.java

```diff
@@ -482,6 +502,49 @@ void reassignPartitions(List<KafkaTopicPartitionState<TopicPartition>> newPartit
 		return new KafkaConsumer<>(kafkaProperties);
 	}
 
+	@VisibleForTesting
+	RateLimiter getRateLimiter() {
+		return rateLimiter;
+	}
+
+	// -----------------------------------------------------------------------
+	// Rate limiting methods
+	// -----------------------------------------------------------------------
+	/**
+	 *
+	 * @param records List of ConsumerRecords.
+	 * @return Total batch size in bytes, including key and value.
+	 */
+	private int getRecordBatchSize(ConsumerRecords<byte[], byte[]> records) {
```

Comment:
It is a little unfortunate that we have to make an additional iteration over the records to get the sizes. Another option is putting the throttling logic in the `AbstractFetcher`.

Reply:
The reason for doing the iteration here was because we needed the record sizes prior to deserialization and I believe the records are already deserialized at the `AbstractFetcher` level?

Reply:
The `AbstractFetcher` has two subclasses - `Kafka09Fetcher` and `Kafka08Fetcher`. The serialization is done inside them. 

`Kafka09Fetcher` is sort of an abstract fetcher for Kafka 0.9+. `Kafka08Fetcher` is more of a legacy. Maybe we can just do that in `Kafka09Fetcher`?